### ISSUE TITLE
Payload Parsing is not disabled using parse: false

### ISSUE THREAD

**Issue by spuddleziz** [hapijs/hapi#3465]
I want to avoid parsing the entire payload for very large file uploads. I did this by setting my route configuration to the following:

`payload : {
  output : "stream",
  parse : false,
  multipart: false
}`

While I do agree that the upload isn't actually getting parsed, req.payload has still been "parsed" in some way. This process reads the entire file into memory and this will consume LOTS of memory for very large uploads.

After lots of messing about I have got this to work by performing the upload using the route extension point onPostAuth and streaming `request.raw.req` to another multipart parser:

https://github.com/pillarjs/multiparty/issues/168

The docs aren't very clear on this - the expected behaviour would be that the payload would be a raw stream, which it is, but its an in memory representation of the original node raw request object wrapped with an object representing the multipart field data. I disabled multipart parsing so I'm unsure why or how this could occur given the configuration.

Thoughts, suggestions and questions welcome!

Node 4.8.0
Hapi 16.1.0
Ubuntu 14.04

**Comment by ubnt-michals** [comment]
We had the same memory issues related to large files uploads. Looks like the payload is still fully loaded to the memory. Here is a quick fix:

```.js
const { Form } = require('multiparty');

const handleFirmwareUpload = (request, reply) => {
  const rawRequest = request.raw.req;
  const uploadPromises = [];
  const form = new Form();

  form.on('error', reply);
  form.on('part', (part) => {
    uploadPromises.push(processFirmwareImage(part));
  });
  form.on('close', () => reply(Promise.all(uploadPromises)));
  form.parse(rawRequest);
};

server.route({
  method: 'POST',
  path: '/v2.1/firmwares',
  config: {
    ext: {
      onPostAuth: {
        method: handleFirmwareUpload,
      },
    },
    payload: {
      output: 'stream',
      maxBytes: 500e6,
      parse: false,
      allow: 'multipart/form-data',
    },
  },
  handler: noop,
});
``` 

I am still not sure how to stop Hapi from 'parsing'/loading the full request.

**Comment by devinivy** [comment]
Could you provide more info?  What do you do with the upload instead of using hapi to read it?  It is true that hapi essentially shouldn't touch `request.raw.req` with the following config,

```json5
    payload: {
      output: 'stream',
      parse: false,
    }
```

**Comment by ubnt-michals** [comment]
@devinivy We have noticed our docker container is running out of memory and we traced to issue to the uploads. It looks like the hapi will load the file or significant portion of it to the memory. Maybe the cause is in 
```.js
 allow: 'multipart/form-data',
```
We used hapi version 16.4.3

Let me know if you need anything else.

**Comment by jeff-kilbride** [comment]
@ubnt-michals  @devinivy 

I traced through this stuff a while back. I believe the issue is in the `stream` implementation. If I remember correctly, hapi reads the entire file into a buffer and then uses that buffer to create a new stream, rather than piping `request.raw.req`.

For my purposes, I am uploading video files to hapi, doing some ffmpeg magic on them, and then pushing them to S3. I was curious whether it would be better to use `output: 'stream'` and stream the files directly to disk, or `output: 'file'`. As it turns out, `file` is much better for me. (I am uploading from multi-part form data, so I have to set `parse: true`)

**Comment by devinivy** [comment]
I would be very interested to know if hapi reads payloads into a buffer even when the payload is streaming and not being parsed.  I would expect that it passes the stream right through.  Can anybody vouch for this with any level of certainty?

**Comment by jeff-kilbride** [comment]
@devinivy  I just traced this again and below is my reading, but I would also be interested in hearing from someone more familiar with it. I've run into more than one post like the OP's claiming that hapi reads the entire upload into memory with parse set to false and output set to stream. However, that's not what I'm seeing in the code.

This is handled by Subtext.

https://github.com/hapijs/hapi/blob/master/lib/route.js#L426

Based on the OP's settings 

`payload : { output : "stream", parse : false, multipart: false }`

It is not being read into a buffer. Setting parse to false puts you in the `raw` method:

https://github.com/hapijs/subtext/blob/master/lib/index.js#L88

and setting output to stream returns `request.raw.req` to the callback supplied to Subtext (as `parsed.payload`):

https://github.com/hapijs/subtext/blob/master/lib/index.js#L210

So, unless I'm reading this wrong, setting parse to false and output to stream should return `request.raw.req` as the payload.

However, in my case, I am using multi-part form data, so I have to set parse to true. With multi-part data, the code goes down a bit of a rabbit-hole into Pez, Nigel, and Vise -- not for the faint of heart 😄  -- as well as using Wreck.read(). Essentially, the entire part is read into a buffer before being piped out:

https://github.com/hapijs/subtext/blob/master/lib/index.js#L129
https://github.com/hapijs/subtext/blob/master/lib/index.js#L392

as also touched on by @mtharrison in the second paragraph of this comment:

https://github.com/hapijs/hapi/issues/3051#issuecomment-185643445

Setting output to `file` streams the data to disk without additional buffering:

https://github.com/hapijs/subtext/blob/master/lib/index.js#L360

Sorry for the length. I hope this makes sense and that someone else can verify that it's correct. It took me several hours to wrap my head around it the first time I traced it (especially the multi-part stuff), but it all came flooding back when I looked at it again a few minutes ago!

**Comment by ubnt-michals** [comment]
> However, in my case, I am using multi-part form data, so I have to set parse to true. With multi-part data, the code goes down a bit of a rabbit-hole into Pez, Nigel, and Vise -- not for the faint of heart :smile: -- as well as using Wreck.read(). Essentially, the entire part is read into a buffer before being piped out..

I think this is the root of the issue. For our config:
```.js
payload: {
      output: 'stream',
      maxBytes: 500e6,
      parse: false,
      allow: 'multipart/form-data',
 },
```
It would still parse the data. Even with `parse: false`! I am not sure if this is by design or something that should be fixed.

**Comment by jeff-kilbride** [comment]
@ubnt-michals  From reading the code, it shouldn't do that. When parse is set to false, it ignores the `multipart/form-data` and just returns the `request.raw.req` stream -- at least, that's what it looks like to me. Maybe I'm missing something...

**Comment by ubnt-michals** [comment]
@jeff-kilbride I'm pretty sure there is still something being buffered. I will explore the issue later today with the latest Hapi version..

**Comment by devinivy** [comment]
Cool, I think we're on the right track.  I agree that it shouldn't buffer in that case, so I will be interested to see where that memory is going!

**Comment by ubnt-michals** [comment]
I can confirm with hapi 16.4.3 the configuration bellow is no longer parsing the request. 

```.js
payload: {
      output: 'stream',
      maxBytes: 500e6,
      parse: false,
      allow: 'multipart/form-data',
 },
```

I see no memory spikes when uploading large files. I have implemented the workaround above for hapi 15.0.3 - looks like somewhere between 15.0.3 and 16.4.3 it got fixed. I think we can close the issue. @devinivy and @jeff-kilbride thank you for your help.

**Comment by hueniverse** [comment]
We should make this clear in the documents. It was somewhere in release notes when the multipart stream feature was added.

Basically, there is no way in node to partially process a multipart stream. If you want to process incoming payload for multiparts, we have to basically read the whole thing at which point we need to simulate a stream for each chunk.

An interface that does what you want will need to defer payload processing for later which is what the proposed solutions above do with multipart (though you could use subtext the same way).

Using temporary files is a good method for large scale. Let's you keep minimal in-memory state and defer processing to another process. It can create lingering crap that needs to be cleaned. 

The right solution here is to create an uploads plugin that is similar to the proxy plugin (h2o2) and then you can use that as a new type of route handler. Someone will probably enjoy creating it (and it can be recommended in the docs).
